### PR TITLE
nixos/grub, nixos/systemd-boot: Replace proprietary Memtest86 with free Memtest86+ on NixOS systems with UEFI

### PR DESCRIPTION
**Pull request by enc0urage**
###### Description of changes

Now that Memtest86+ supports UEFI and the 64-bit Linux boot handover protocol, we can use that on UEFI systems instead of falling back to the proprietary Passmark Memtest86. 

GRUB2 no longer needs any special handling for booting memtest86+ on UEFI systems, as it can be booted both using the 64-bit Linux handover protocol on UEFI and 16-bit Linux handover on legacy systems. 

This allows for a better implementation than #225616 for GRUB2, because we don't need any special handling for UEFI anymore, it can use the same entry/binary as legacy boot, and still accept boot params in UEFI.

On systemd-boot we need to use the new UEFI binary as in #225616.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/loader/grub/memtest.nix [hunk 1]
```diff
@@ -1,35 +1,28 @@
...
-          `true`.
+          Make Memtest86+, a memory testing program, available from the GRUB
+          boot menu.
         '';
       };
 
       params = mkOption {
         default = [];
-        example = [ "console=ttyS0,115200" ];
+        example = ["console=ttyS0,115200"];
```
Comment (RaitoBezarius) [inline nixos/modules/system/boot/loader/grub/memtest.nix:25]: Unrelated change

File: nixos/modules/system/boot/loader/grub/memtest.nix [hunk 2]
```diff
@@ -63,38 +58,16 @@ in
...
+        menuentry "Memtest86+" {
+          linux @bootRoot@/memtest.bin ${toString cfg.params}
         }
-      ];
-
-      boot.loader.grub.extraFiles = {
-        "memtest86.efi" = "${pkgs.memtest86-efi}/BOOTX64.efi";
-      };
+      ''
+      else throw "Memtest86+ is not supported with GRUB 1.";
```
Comment (RaitoBezarius) [inline nixos/modules/system/boot/loader/grub/memtest.nix:69]: This throw will be unneeded if #167204 is merged.

File: nixos/modules/system/boot/loader/grub/memtest.nix [hunk 3]
```diff
@@ -63,38 +58,16 @@ in
...
-              linux16 @bootRoot@/memtest.bin ${toString cfg.params}
-            }
-          ''
-        else
-          throw "Memtest86+ is not supported with GRUB 1.";
-
-      boot.loader.grub.extraFiles."memtest.bin" = "${memtest86}/memtest.bin";
-    })
-  ];
+    boot.loader.grub.extraFiles."memtest.bin" = "${memtest86}/memtest.bin";
```
Comment (RaitoBezarius) [inline nixos/modules/system/boot/loader/grub/memtest.nix:71]: Does this handle UEFI?

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix [hunk 4]
```diff
@@ -280,7 +277,7 @@ in {
       # be updated to use the 32-bit EFI app on 32-bit systems.  The 32-bit EFI
       # app filename is BOOTIA32.efi.
       (mkIf cfg.memtest86.enable {
-        "efi/memtest86/BOOTX64.efi" = "${pkgs.memtest86-efi}/BOOTX64.efi";
+        "efi/memtest86/memtest.efi" = "${pkgs.memtest86plus.efi}";
```
Comment (RaitoBezarius) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:280]: It seems like the package expose a `.efi` passthrough, it would be good to have a symmetric code for GRUB by having a `legacy` and `hybrid` passthrough for the other possibilities.

**Review by RaitoBezarius — CHANGES_REQUESTED** [review]
Still reading this change, please rename the PR title and commits (if not done already) according to the CONTRIBUTING guide.
I notice we do not have GRUB tests for memtest, it would be interesting to add them as GRUB is known to break a lot.

**Comment by RaitoBezarius** [comment]
Use only the most important one, or wildcards or separate them by comma

File: nixos/modules/system/boot/loader/grub/memtest.nix (hunk 2 above)
Comment (RaitoBezarius) [reply nixos/modules/system/boot/loader/grub/memtest.nix:69]: The throw is unneeded, please rebase

File: nixos/modules/system/boot/loader/grub/memtest.nix (hunk 3 above)
Comment (RaitoBezarius) [reply nixos/modules/system/boot/loader/grub/memtest.nix:71]: Nice

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix (hunk 4 above)
Comment (RaitoBezarius) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:280]: Got it

**Comment by RaitoBezarius** [comment]
@ofborg test systemd-boot.memtest86

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix [hunk 5]
```diff
@@ -194,7 +191,7 @@ in {
       example = literalExpression ''
         { "memtest86.conf" = '''
           title MemTest86
```
Comment (andersk) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:193]: Title should be updated to Memtest86+.

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix (hunk 4 above)
Comment (samueldr) [reply nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:277]: (Note about resolution, discussed about it in the matrix room, and it is not necessary to deal with the 32 bit path since it is intrinsically handled by the i686 system being built)

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix [hunk 6]
```diff
@@ -147,11 +147,8 @@ in {
         default = false;
         type = types.bool;
         description = lib.mdDoc ''
-          Make MemTest86 available from the systemd-boot menu. MemTest86 is a
-          program for testing memory.  MemTest86 is an unfree program, so
-          this requires `allowUnfree` to be set to
-          `true`.
-        '';
+          Make MemTest86+ available from the systemd-boot menu. MemTest86+ is a
+          program for testing memory.'';
```
Comment (samueldr) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot.nix:151]: Trivial nit, if you can't handle it (time constraint, ate too much, any reason) do tell and I'll handle that.

```suggestion
          Make MemTest86+ available from the systemd-boot menu. MemTest86+ is a
          program for testing memory.
        '';
```

**Comment by samueldr** [comment]
I still need to test it, just in case, but other than the nit it seems alright from reading the code.

**Comment by ajs124** [comment]
Sadly memtest86+ does not support aarch64 and seems like they're not planning on adding support either (https://github.com/memtest86plus/memtest86plus/issues/285, https://github.com/memtest86plus/memtest86plus/issues/147), so this is a breaking change on that platform.
